src/Fixer/Import/NoUnusedImportsFixer.php

```diff
@@ -81,62 +81,26 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
             return;
         }
 
-        $useDeclarations = $this->getNamespaceUseDeclarations($tokens, $useDeclarationsIndexes);
-        $namespaceDeclarations = $this->getNamespaceDeclarations($tokens);
-        $contentWithoutUseDeclarations = $this->generateCodeWithoutPartials($tokens, array_merge($namespaceDeclarations, $useDeclarations));
-        $useUsages = $this->detectUseUsages($contentWithoutUseDeclarations, $useDeclarations);
+        foreach ($this->getNamespaces($tokens) as $namespace) {
+            $useDeclarations = $this->getNamespaceUseDeclarations($tokens, $namespace, $useDeclarationsIndexes);
 
-        $this->removeUnusedUseDeclarations($tokens, $useDeclarations, $useUsages);
-        $this->removeUsesInSameNamespace($tokens, $useDeclarations, $namespaceDeclarations);
-    }
-
-    /**
-     * @param string $content
-     * @param array  $useDeclarations
-     *
-     * @return array
-     */
-    private function detectUseUsages($content, array $useDeclarations)
-    {
-        $usages = array();
-
-        foreach ($useDeclarations as $shortName => $useDeclaration) {
-            $usages[$shortName] = (bool) preg_match('/(?<![\$\\\\])(?<!->)\b'.preg_quote($shortName).'\b/i', $content);
-        }
-
-        return $usages;
-    }
+            $usagesSearchIgnoredIndexes = array();
 
-    /**
-     * @param Tokens $tokens
-     * @param array  $partials
-     *
-     * @return string
-     */
-    private function generateCodeWithoutPartials(Tokens $tokens, array $partials)
-    {
-        $content = '';
-
-        foreach ($tokens as $index => $token) {
-            $allowToAppend = true;
-
-            foreach ($partials as $partial) {
-                if ($partial['start'] <= $index && $index <= $partial['end']) {
-                    $allowToAppend = false;
+            foreach ($useDeclarations as $useDeclaration) {
+                $usagesSearchIgnoredIndexes[$useDeclaration['start']] = $useDeclaration['end'];
+            }
 
-                    break;
+            foreach ($useDeclarations as $shortName => $useDeclaration) {
+                if (!$this->importIsUsed($tokens, $namespace, $usagesSearchIgnoredIndexes, $shortName)) {
+                    $this->removeUseDeclaration($tokens, $useDeclaration);
                 }
             }
 
-            if ($allowToAppend) {
-                $content .= $token->getContent();
-            }
+            $this->removeUsesInSameNamespace($tokens, $useDeclarations, $namespace);
         }
-
-        return $content;
     }
 
-    private function getNamespaceDeclarations(Tokens $tokens)
+    private function getNamespaces(Tokens $tokens)
```

Comment:
there seem to be a lot of interest in getting NS tokens in PR's ATM;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3223/files#diff-b8963a9ef67a6665ad5319f29b9a3e24R366
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3276/files#diff-1e2d90e49bf3d7b2bd4d5088aa553d5bR1

maybe we can unify to one solution (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3276 seems most re-usable)

Reply:
Makes sense, shall I do this in this PR?

Reply:
sounds good to me :)